### ISSUE TITLE
JS/WS: Nonce not increasing - after random time - error

### ISSUE THREAD

**Issue by ogoossens** [BitMEX/api-connectors#91]
**Hi,**

I'm running an authenticated WS connection build in JavaScript (Node.js) to see my open position stats.

What happens - on a totally irregular basis, I get an error about nonce not increasing. 

I know the idea about nonce but I have no idea why this error occurs since:

- **Im running a single app (no multiple use of API keys)**
- its total random
- it can run without any errors for 10 min, or it can get this error after 20-30 secs

Why does this happen, and how to fix it please?

**P.S.:** I have read about read about rewriting the nonce but I cant imagine how that could fix my problem, since after I restart the app it starts with no issues... hmm?

**P.P.S.:** Beside that Im just logged in via chrome to also check my positions, but that should not abuse my API key, right?

**Thank you**

**Oliver**

**Comment by ryanfox** [comment]
Thanks for the report. The only way you should get a non-increasing nonce is if messages are received out of order. The nonce is generated when the message is signed, based on `Date.now()` - if it is not modified, it should be very difficult to get a lower nonce.

Is your application doing anything in particular when the error occurs?

Chrome should not conflict with your API keys.

**Comment by ogoossens** [comment]
Great reply, thank you for caring :) - yes I thought browser will not affect this. My app? Nothing special just listening to positions, very very basic and once it breaks I get an error saying something like "nonve not increasing bla bla ... shot down restart the app manually" ... so what I did I got an npm package called FOREVER and once it sees my app crashing, it restarts the app - and everythign goes as before ... 

Sometimes its minutes ok, sometimes 2-3 hours, but then time ti time this error ... than restart and we continue. I mean for my purpose lets say its ok, because my restart takes 5-8 secs, but should I depend on it more I would be not happy :-/ but as said, no idea what causes this.. but when I think of it this is important : 

The error says nonce not increasing - and I dont get a SMALLER one but EXACT same one as the last one, as if the nonce is RE-used or something like this...

Its not like the last nonce is 150 and I got 149 - that could be out of order, but last nonce is 150 and I get 150 again, thats the issue :)

**Comment by ryanfox** [comment]
Have you checked that the actual nonce is the same? That's just the message when a nonce arrives with less or equal value.

That would indicate a different sort of problem.

**Comment by ogoossens** [comment]
The error says it .. I could reproduce it and send you a screenshot but it says something like : 

Nonce not increasing. Current nonce XXXXXXXXX546 last nonce XXXXXXXXXXXX546 (same number)

Thats what bugged my head ...

**Comment by ryanfox** [comment]
In that case, it seems like the client is sending the same request multiple times.

A log of the messages sent would be helpful to diagnose what is happening. On the other hand, if simply restarting the application is easier, that is a workaround as well.

**Comment by ogoossens** [comment]
Yes, it is a workaround that works for me :) but the question is why it happens in the first place when I dont do anything special.. Okay then, thank you for the response. Cheers

**Comment by annugaurish** [comment]
I have solved this problem by making changes in createSocket.js
Added following code in the createSocket function

wsClient.onreconnect = function() {
         wsClient.url = makeEndpoint(options);
		
	debug('Reconnecting to BitMEX at ', wsClient.url);
};

and in ReconnectionSocket.js added the following line in reconnect prototype
   this.onreconnect();

this is to be added before this.open(this.url);

**Comment by ogoossens** [comment]
You Sir, are my hero. Thank you :-)

**Comment by annugaurish** [comment]
You are most welcome. BTW I am mam, not sir :P

**Comment by pabx06** [comment]
i relay dont get it. why changes are not merged up with the connector ?
each time app is deployed using npm install you would have to edit libs by hands the connector
